File: lambda-simple/handler.js
/**
 * Simple WebSocket Handler for AWS API Gateway
 * Handles all routes in a single Lambda function
 */

const { ApiGatewayManagementApiClient, PostToConnectionCommand } = require('@aws-sdk/client-apigatewaymanagementapi');

/**
 * Create API Gateway Management API client
 */
function getApiClient(domainName, stage) {
    const endpoint = `https://${domainName}/${stage}`;
    return new ApiGatewayManagementApiClient({
        endpoint: endpoint
    });
}

/**
 * Send message to a specific connection
 */
async function sendToConnection(client, connectionId, data) {
    try {
        const command = new PostToConnectionCommand({
            ConnectionId: connectionId,
            Data: JSON.stringify(data)
        });
        
        await client.send(command);
        return true;
    } catch (error) {
        if (error.name === 'GoneException') {
            console.log(`Connection ${connectionId} is gone`);
        } else {
            console.error(`Error sending to ${connectionId}:`, error);
        }
        return false;
    }
}

/**
 * Main Lambda handler for all WebSocket events
 */
exports.handler = async (event, context) => {
    const routeKey = event.requestContext.routeKey;
    const connectionId = event.requestContext.connectionId;
    const domainName = event.requestContext.domainName;
    const stage = event.requestContext.stage;
    
    console.log(`Route: ${routeKey}, Connection: ${connectionId}`);
    
    // Create API client
    const apiClient = getApiClient(domainName, stage);
    
    try {
        // Handle routes
        if (routeKey === '$connect') {
            // New connection
            console.log(`Client connected: ${connectionId}`);
            // Just accept the connection
            // We'll send the connection ID when client first subscribes
            return { statusCode: 200 };
        }
        
        else if (routeKey === '$disconnect') {
            // Connection closed
            console.log(`Client disconnected: ${connectionId}`);
            return { statusCode: 200 };
        }
        
        else if (routeKey === '$default') {
            // Handle all messages
            const body = JSON.parse(event.body || '{}');
            const action = body.action || 'unknown';
            
            console.log(`Action: ${action}, Data:`, body);
            
            if (action === 'subscribe') {
                // Subscribe to channel
                const channel = body.channel || 'default';
                
                // First, send the connection ID to the client
                await sendToConnection(apiClient, connectionId, {
                    type: 'connection_id',
                    connection_id: connectionId
                });
                
                // Then send subscription confirmation
                const response = {
                    type: 'subscription_succeeded',
                    channel: channel
                };
                await sendToConnection(apiClient, connectionId, response);
            }
            
            else if (action === 'unsubscribe') {
                // Unsubscribe from channel
                const channel = body.channel || 'default';
                const response = {
                    type: 'unsubscribe_succeeded',
                    channel: channel
                };
                await sendToConnection(apiClient, connectionId, response);
            }
            
            else if (action === 'message') {
                // Echo message back (since we don't have connection list)
                const response = {
                    type: 'message',
                    data: body.data || {}
                };
                await sendToConnection(apiClient, connectionId, response);
            }
            
            else {
                // Unknown action
                const response = {
                    type: 'error',
                    message: `Unknown action: ${action}`
                };
                await sendToConnection(apiClient, connectionId, response);
            }
            
            return { statusCode: 200 };
        }
        
        // Unknown route
        return { statusCode: 404 };
        
    } catch (error) {
        console.error('Error processing message:', error);
        return { statusCode: 500 };
    }
};

File: lambda-simple/handler.py
"""
Simple WebSocket Handler for AWS API Gateway
Handles all routes in a single Lambda function
"""

import json
import boto3
import os

# API Gateway Management API client (initialized per request)
def get_api_client(domain_name, stage):
    """Create API Gateway Management API client"""
    endpoint_url = f"https://{domain_name}/{stage}"
    return boto3.client('apigatewaymanagementapi', endpoint_url=endpoint_url)


def send_to_connection(client, connection_id, data):
    """Send message to a specific connection"""
    try:
        client.post_to_connection(
            ConnectionId=connection_id,
            Data=json.dumps(data).encode('utf-8')
        )
        return True
    except client.exceptions.GoneException:
        print(f"Connection {connection_id} is gone")
        return False
    except Exception as e:
        print(f"Error sending to {connection_id}: {e}")
        return False


def broadcast_to_all(client, sender_id, data, exclude_sender=False):
    """
    Broadcast message to all connected clients
    
    Note: API Gateway doesn't provide a way to list all connections.
    In production, you'd need to track connections in DynamoDB or Redis.
    For this simple version, we'll just acknowledge receipt.
    """
    # Since we can't list connections without storage,
    # just echo back to sender for this simple implementation
    if not exclude_sender:
        send_to_connection(client, sender_id, data)
    
    # In a real implementation with connection tracking:
    # connections = get_all_connections_from_storage()
    # for conn_id in connections:
    #     if exclude_sender and conn_id == sender_id:
    #         continue
    #     send_to_connection(client, conn_id, data)


def lambda_handler(event, context):
    """
    Main handler for all WebSocket events
    """
    
    route_key = event['requestContext']['routeKey']
    connection_id = event['requestContext']['connectionId']
    domain_name = event['requestContext']['domainName']
    stage = event['requestContext']['stage']
    
    print(f"Route: {route_key}, Connection: {connection_id}")
    
    # Create API client
    api_client = get_api_client(domain_name, stage)
    
    # Handle routes
    if route_key == '$connect':
        # New connection
        print(f"Client connected: {connection_id}")
        return {'statusCode': 200}
    
    elif route_key == '$disconnect':
        # Connection closed
        print(f"Client disconnected: {connection_id}")
        return {'statusCode': 200}
    
    elif route_key == '$default':
        # Handle all messages
        try:
            body = json.loads(event.get('body', '{}'))
            action = body.get('action', 'unknown')
            
            print(f"Action: {action}, Data: {body}")
            
            if action == 'subscribe':
                # Subscribe to channel
                channel = body.get('channel', 'default')
                response = {
                    'type': 'subscription_succeeded',
                    'channel': channel
                }
                send_to_connection(api_client, connection_id, response)
                
            elif action == 'message':
                # Broadcast message
                # Echo back to sender (since we don't have connection list)
                response = {
                    'type': 'message',
                    'data': body.get('data', {})
                }
                send_to_connection(api_client, connection_id, response)
            
            else:
                # Unknown action
                response = {
                    'type': 'error',
                    'message': f'Unknown action: {action}'
                }
                send_to_connection(api_client, connection_id, response)
            
            return {'statusCode': 200}
            
        except Exception as e:
            print(f"Error processing message: {e}")
            return {'statusCode': 500}
    
    return {'statusCode': 404}
